### PR TITLE
Fixed #802 - Added custom volume preference for "My Reminder"

### LINKED ISSUE #802: Add custom volume for "My reminders" notifications
**Summary:** 

Thanks to @bridgette via https://github.com/OneBusAway/onebusaway-android/pull/801, we now have custom controls over notification sounds and vibration for the "My reminders" feature.

Here's an additional request for notification customization from a user, via Ideascale - http://onebusaway.ideascale.com/a/dtd/Notifications/537373-8715:

>I like the "set alarm" notification, but an additional customization option would be nice so that I can hear the notification. My default tone is not very loud (on purpose), I would set this app to have a louder notification tone - if I could set that.

**Steps to reproduce:** 

1. Tap on arrival time and choose "Set a reminder", and set a reminder to go off for an arrival
1. Listen to the notification volume when it's triggered

Then, go to app "Settings".

**Expected behavior:** 

Give me control over the OBA notification volume independent of system notification volume (most likely in the Settings menu under "Notification" section).

**Observed behavior:** 

There is no way to independently control OBA notification volume

**Device and Android version:** 

N/A

### PR DESCRIPTION
**Pull request by sebesti0n**
Fixed #802 
Preview:

https://github.com/OneBusAway/onebusaway-android/assets/114914556/26a0e7cb-c9b7-48ed-8751-b2dcb0364ff4

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Apply the `AndroidStyle.xml` style template to your code in Android Studio.

- [x] Run the unit tests with `gradlew connectedObaGoogleDebugAndroidTest` to make sure you didn't break anything

- [x] If you have multiple commits please combine them into one commit by squashing them for the initial submission of the pull request.  When addressing comments on a pull request, please push a new commit per comment when possible (reviewers will squash and merge using GitHub merge tool)

### CODE REVIEW

File: onebusaway-android/src/main/java/org/onebusaway/android/ui/PreferencesActivity.java [hunk 1]
```diff
@@ -121,6 +126,11 @@ public class PreferencesActivity extends PreferenceActivity
 
     ListPreference mThemePref;
 
+
```
Comment (aaronbrethorst) [inline onebusaway-android/src/main/java/org/onebusaway/android/ui/PreferencesActivity.java:129]: please watch your newlines in between instance variables. one blank newline is fine; two looks strange.

File: onebusaway-android/src/main/java/org/onebusaway/android/ui/PreferencesActivity.java [hunk 2]
```diff
@@ -185,7 +195,11 @@ public void onCreate(Bundle savedInstanceState) {
         mAboutPref = findPreference(getString(R.string.preferences_key_about));
         mAboutPref.setOnPreferenceClickListener(this);
 
-        SharedPreferences settings = Application.getPrefs();
+        mVolumePref = findPreference("reminder_volume");
```
Comment (aaronbrethorst) [inline onebusaway-android/src/main/java/org/onebusaway/android/ui/PreferencesActivity.java:198]: extract to resources like the other preferences keys.

File: onebusaway-android/src/main/java/org/onebusaway/android/ui/PreferencesActivity.java [hunk 3]
```diff
@@ -366,10 +380,52 @@ public boolean onPreferenceClick(Preference pref) {
...
         }
         return true;
     }
 
+    /*
+    * Dialog to set the volume for reminder using seekbar
+    * save volume in shared Preferences
+    */
+    private void reminderVolumeDialog() {
+        final int progress = settings.getInt("reminder_volume",100);
```
Comment (aaronbrethorst) [inline onebusaway-android/src/main/java/org/onebusaway/android/ui/PreferencesActivity.java:394]: same here as above with the preferences key

File: onebusaway-android/src/main/java/org/onebusaway/android/ui/PreferencesActivity.java [hunk 4]
```diff
@@ -366,10 +380,52 @@ public boolean onPreferenceClick(Preference pref) {
...
+        LayoutInflater inflater = getLayoutInflater();
+        View customLayout = inflater.inflate(R.layout.reminder_volume_adjuster, null);
+        SeekBar seekBar = customLayout.findViewById(R.id.seekBar);
+        AlertDialog.Builder builder = new AlertDialog.Builder(this)
+                .setView(customLayout)
+                .setPositiveButton(R.string.save,
+                        (dialog, which)->{
+                                setReminderVolume(seekBar.getProgress());
```
Comment (aaronbrethorst) [inline onebusaway-android/src/main/java/org/onebusaway/android/ui/PreferencesActivity.java:402]: there's an extra tab on this line, please remove it.

File: onebusaway-android/src/main/java/org/onebusaway/android/ui/PreferencesActivity.java [hunk 5]
```diff
@@ -366,10 +380,52 @@ public boolean onPreferenceClick(Preference pref) {
...
+                        }
+                );
+        seekBar.setProgress(progress);
+        AlertDialog dialog = builder.create();
+        dialog.show();
+    }
+
+    private void setReminderVolume(int progress) {
+        AudioManager audioManager = null;
+        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) {
```
Comment (aaronbrethorst) [inline onebusaway-android/src/main/java/org/onebusaway/android/ui/PreferencesActivity.java:416]: if this feature only works on Android M and above, then it should only be displayed to users on Android M and above.

File: onebusaway-android/src/main/java/org/onebusaway/android/ui/PreferencesActivity.java [hunk 6]
```diff
@@ -366,10 +380,52 @@ public boolean onPreferenceClick(Preference pref) {
...
+            audioManager = (AudioManager) getSystemService(Context.AUDIO_SERVICE);
+        }
+        if (audioManager != null) {
```
Comment (aaronbrethorst) [inline onebusaway-android/src/main/java/org/onebusaway/android/ui/PreferencesActivity.java:419]: I recommend restructuring this method to use early returns.

```java
private void setReminderVolume(int progress) {
        AudioManager audioManager = null;
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) {
            audioManager = (AudioManager) getSystemService(Context.AUDIO_SERVICE);
        }
        
        if (audioManager == null) {
            return;
        }

        int maxVolume = audioManager.getStreamMaxVolume(AudioManager.STREAM_ALARM);
        int scaledVolume = (int) (progress / 100f * maxVolume);
        audioManager.setStreamVolume(AudioManager.STREAM_ALARM, scaledVolume, 0);
        SharedPreferences.Editor editor = PreferenceManager.getDefaultSharedPreferences(this).edit();
        editor.putInt("reminder_volume", progress);
        editor.apply();
    }
```

File: onebusaway-android/src/main/java/org/onebusaway/android/ui/PreferencesActivity.java [hunk 7]
```diff
@@ -366,10 +380,52 @@ public boolean onPreferenceClick(Preference pref) {
...
+            int maxVolume = audioManager.getStreamMaxVolume(AudioManager.STREAM_ALARM);
+            int scaledVolume = (int) (progress / 100f * maxVolume);
+            audioManager.setStreamVolume(AudioManager.STREAM_ALARM, scaledVolume, 0);
+            SharedPreferences.Editor editor = PreferenceManager.getDefaultSharedPreferences(this).edit();
+            editor.putInt("reminder_volume", progress);
```
Comment (aaronbrethorst) [inline onebusaway-android/src/main/java/org/onebusaway/android/ui/PreferencesActivity.java:424]: resource string again

File: onebusaway-android/src/main/res/values-es/strings.xml [hunk 8]
```diff
@@ -1077,4 +1077,8 @@
     <string name="trip_info_grant_exact_alarms_permission_message">Debes otorgar permiso a la aplicación para programar un recordatorio.</string>
     <string name="trip_info_grant_exact_alarms_permission_positive_button">Otorgar permiso</string>
     <string name="trip_info_grant_exact_alarms_permission_negative_button">Cancelar</string>
+
+    <!--    reminder volume adjuster layout-->
```
Comment (aaronbrethorst) [inline onebusaway-android/src/main/res/values-es/strings.xml:1081]: please watch your spaces:

```
<!-- reminder volume adjuster layout -->
```

File: onebusaway-android/src/main/res/values-es/strings.xml [hunk 9]
```diff
@@ -1077,4 +1077,8 @@
...
+    <string name="reminder_volume_adjuster_title">volumen</string>
```
Comment (aaronbrethorst) [inline onebusaway-android/src/main/res/values-es/strings.xml:1082]: Capitalize the title

File: onebusaway-android/src/main/res/values-es/strings.xml [hunk 10]
```diff
@@ -1077,4 +1077,8 @@
...
+    <string name="save">guardar</string>
```
Comment (aaronbrethorst) [inline onebusaway-android/src/main/res/values-es/strings.xml:1083]: Change the string key from `save` to `reminder_volume_adjuster_save` and capitalize

File: onebusaway-android/src/main/res/values-fi/strings.xml [hunk 11]
```diff
@@ -672,4 +672,8 @@
     <string name="trip_info_grant_exact_alarms_permission_message">Sinun täytyy antaa sovellukselle lupa muistutuksen asettamiseen.</string>
     <string name="trip_info_grant_exact_alarms_permission_positive_button">Myönnä lupa</string>
     <string name="trip_info_grant_exact_alarms_permission_negative_button">Peruuta</string>
+
+    <!--    reminder volume adjuster layout-->
```
Comment (aaronbrethorst) [inline onebusaway-android/src/main/res/values-fi/strings.xml:676]: same as above

File: onebusaway-android/src/main/res/values-fi/strings.xml [hunk 12]
```diff
@@ -672,4 +672,8 @@
...
+    <string name="reminder_volume_adjuster_title">tilavuus</string>
+    <string name="save">"tallentaa "</string>
```
Comment (aaronbrethorst) [inline onebusaway-android/src/main/res/values-fi/strings.xml:678]: also, there shouldn't be double quotes inside the XML element here.

File: onebusaway-android/src/main/res/values-it/strings.xml [hunk 13]
```diff
@@ -988,4 +988,8 @@
     <string name="trip_info_grant_exact_alarms_permission_message">Devi concedere all\'app il permesso di pianificare un promemoria.</string>
     <string name="trip_info_grant_exact_alarms_permission_positive_button">Concedi il permesso</string>
     <string name="trip_info_grant_exact_alarms_permission_negative_button">Annulla</string>
+
+    <!--    reminder volume adjuster layout-->
```
Comment (aaronbrethorst) [inline onebusaway-android/src/main/res/values-it/strings.xml:992]: same as above, including the double quotes inside the XML element

File: onebusaway-android/src/main/res/values-pl/strings.xml [hunk 14]
```diff
@@ -613,6 +613,10 @@
 
     <!-- About -->
     <string name="title_activity_about">O KiedyBus</string>
+    <!--    reminder volume adjuster layout-->
```
Comment (aaronbrethorst) [inline onebusaway-android/src/main/res/values-pl/strings.xml:616]: Same as above, including double quotes. Also, there are several space characters on the `save` string that need to be removed.

File: onebusaway-android/src/main/res/values/strings.xml [hunk 15]
```diff
@@ -1227,4 +1227,8 @@
     <string name="travel_behavior_dialog_opt_out_message">If you opt out of the research study, we won\'t receive any more information about your travel behavior to help improve public transit.</string>
     <string name="cancel">Cancel</string>
     <string name="my_option_clear_starred_routes">Remove all starred routes</string>
+
+<!--    reminder volume adjuster layout-->
```
Comment (aaronbrethorst) [inline onebusaway-android/src/main/res/values/strings.xml:1231]: indent, spaces, save key, capitalize Save.

File: onebusaway-android/src/main/res/xml/preferences.xml [hunk 16]
```diff
@@ -75,6 +75,13 @@
                 android:key="@string/preference_key_app_theme"
                 android:title="@string/preferences_app_theme_title" />
     </PreferenceCategory>
+    <PreferenceCategory android:title="Reminder"
+        android:key="reminder">
+            <Preference
+                android:key="reminder_volume"
```
Comment (aaronbrethorst) [inline onebusaway-android/src/main/res/xml/preferences.xml:81]: resource key

**Review by aaronbrethorst — CHANGES_REQUESTED** [review]
I see a handful of code changes. Mostly nitpicks. Let me know when this is ready for re-review and thanks for the feature!

**Comment by sebesti0n** [comment]
> I see a handful of code changes. Mostly nitpicks. Let me know when this is ready for re-review and thanks for the feature!

Applied all changes

**Review by aaronbrethorst — CHANGES_REQUESTED** [review]
Code looks good, however I see that this feature is controlling the system-wide alarm volume, not just the volume of OBA-generated alerts. That seems crucial to me, because otherwise we should simply defer to the built-in control for changing volume.

**Comment by sebesti0n** [comment]
> Code looks good, however I see that this feature is controlling the system-wide alarm volume, not just the volume of OBA-generated alerts. That seems crucial to me, because otherwise we should simply defer to the built-in control for changing volume.

Thank you for correcting me. I researched the issue and found that there is no built-in functionality for changing the alarm volume of individual applications. Therefore, I think this could be implemented by decreasing the system volume just before the reminder, and after the reminder is over, returning the system volume to its previous level.

**Pull request closed without merging**